Basic playwright electron support

xai opened the pull request:
<!--
Thank you for your Pull Request. Please provide a description and review
the requirements below.

Contributors guide: https://github.com/theia-ide/theia/blob/master/CONTRIBUTING.md

Note: Security vulnerabilities should not be disclosed on GitHub, through a PR or any
other means. See SECURITY.md at the root of this repository, to learn how to report
vulnerabilities.
-->

#### What it does
<!-- Include relevant issues and describe how they are addressed. -->
Allows to write tests against an Electron app by starting the Electron app first and then attaching Playwright to it. Also, it solves the workspace problem by creating it upfront and pointing to it on app start.

Because playwright cannot access native UI elements, we introduce an optional environment variable `THEIA_ELECTRON_DISABLE_NATIVE_ELEMENTS` that, when set to `1`, enforces the rendering of HTML-based menus and file chooser dialogs, which can be accessed by playwright.

Playwright also requires us to disable early rendering of the window, which we also do via an environment variable (`THEIA_ELECTRON_NO_EARLY_WINDOW`).

A unified TheiaAppLoader is introduced that can be used in both browser and electron tests.
All existing tests are adjusted to use the new AppLoader and can be run in Electron if the environment variable `USE_ELECTRON` is set to `true`.
For convenience, a yarn target `ui-tests-electron` has been added to the playwright `package.json`.
Furthermore, the target `ui-tests-ci` has been added (runs all playwright tests in the browser and in Electron) and the Github workflow has been adjusted.

Contributed on behalf of STMicroelectronics

#### How to test
<!-- Explain how a reviewer can reproduce a bug, test new functionality, or verify performance improvements. -->
1. Build the electron example as usual
  `yarn electron build`
2. Verify that electron is working
  `yarn electron start`
3. Run the electron playwright tests:
   ```
   cd examples/playwright
   yarn build
   yarn ui-tests-electron
   ```
   The testsuite automatically sets all required environment variables, so there is no need to do so manually.

Note that, #12405 still prevents running the playwright tests on Windows or MacOS, which means testing this PR currently requires Linux.

I noticed some flakiness in both browser and electron playwright tests (especially the ones for text-editor, terminal, and outline-view), but I could reproduce these issues on master, so in my opinion, they are not specific to changes introduced in this PR.

#### Review checklist

- [x] As an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- As a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

tsmaeder commented:
@xai could you rebase your changes on latest master to make sure the ubuntu test failures are not due to the old flakiness of the test suite?

tsmaeder commented on examples/playwright/src/tests/theia-explorer-view.test.ts:
```diff
@@ -14,23 +14,30 @@
-let explorer: TheiaExplorerView;
 
+// the tests in this file reuse a page to run faster and thus are executed serially
+test.describe.configure({ mode: 'serial' });
```
Is this fixing a problem you found or is it related to electron? Also: is this a per-file setting or what's the scope?

xai replied:
That is required for the stability of the testsuite, which is improved if the tests are run sequentially. When the tests are run in parallel, we experience a LOT more flakiness. Scope is per file to my knowledge.

tsmaeder commented on packages/core/src/electron-main/electron-main-application.ts:
```diff
@@ -197,7 +197,8 @@ export class ElectronMainApplication {
     async start(config: FrontendApplicationConfig): Promise<void> {
-        this.useNativeWindowFrame = this.getTitleBarStyle(config) === 'native';
+        const args = this.processArgv.getProcessArgvWithoutBin(process.argv);
+        this.useNativeWindowFrame = this.getTitleBarStyle(config) === 'native' && !args.includes('--no-native-window-frame');
```
Would it make sense to make this a theia `CliContribution` that we can include into apps to be tested?

inlann commented:
Hello:

> Because playwright cannot access native UI elements, we introduce an optional command line switch --no-native-window-frame that enforces the rendering of HTML-based menus, which can be accessed by playwright.

HTML-based menus don't seem to work on MacOS:

<img width="1926" alt="image" src="https://user-images.githubusercontent.com/33852855/222677128-72f7b1c6-dd75-45e8-ae1e-1d44aaf58f4e.png">

So, the test failed 😢 :

```bash
mac@Titans-iMac playwright % yarn ui-tests -g "electron"
yarn run v1.22.19
$ yarn && playwright test --config=./configs/playwright.config.ts -g electron
[1/5] 🔍  Validating package.json...
[2/5] 🔍  Resolving packages...
success Already up-to-date.
$ yarn clean && yarn build
$ rimraf lib *.tsbuildinfo
$ tsc --incremental && npx playwright install chromium

Running 7 tests using 1 worker

  ✓  1 …s:34:9 › Theia Electron Application › should load and show main content panel (15ms)
  ✘  2 ../../src/tests/theia-electron-app.test.ts:38:9 › Theia Electron Application › open about dialog using menu (30.0s)
  ✘  3 ../../src/tests/theia-electron-app.test.ts:47:9 › Theia Electron Application › toggle explorer view using menu (30.0s)
  ✓  4 ../../src/tests/theia-electron-app.test.ts:55:9 › Theia Electron Application › open quick command palette (5.1s)
  ✓  5 ../../src/tests/theia-electron-app.test.ts:62:9 › Theia Electron Application › open about dialog using command (5.8s)
  ✓  6 ../../src/tests/theia-electron-app.test.ts:73:9 › Theia Electron Application › select all using command (11.4s)
  ✓  7 …/../src/tests/theia-electron-app.test.ts:80:9 › Theia Electron Application › toggle explorer view using command (7.1s)


  1) ../../src/tests/theia-electron-app.test.ts:38:9 › Theia Electron Application › open about dialog using menu 

    page.waitForSelector: Timeout 30000ms exceeded.
    =========================== logs ===========================
    waiting for locator('#theia\\:menubar .p-MenuBar-itemLabel').locator('text=Help') to be visible
    ============================================================

       at ../../src/theia-main-menu.ts:47

      45 |
      46 |     protected menuBarItem(label = ''): Promise<ElementHandle<SVGElement | HTMLElement> | null> {
    > 47 |         return this.page.waitForSelector(this.menuBarItemSelector(label));
         |                          ^
      48 |     }
      49 |
      50 |     protected menuBarItemSelector(label = ''): string {

        at TheiaMenuBar.menuBarItem (/Users/mac/Work/eclipsesource_theia/examples/playwright/src/theia-main-menu.ts:47:26)
        at TheiaMenuBar.openMenu (/Users/mac/Work/eclipsesource_theia/examples/playwright/src/theia-main-menu.ts:30:40)
        at /Users/mac/Work/eclipsesource_theia/examples/playwright/src/tests/theia-electron-app.test.ts:40:34

  2) ../../src/tests/theia-electron-app.test.ts:47:9 › Theia Electron Application › toggle explorer view using menu 

    page.waitForSelector: Timeout 30000ms exceeded.
    =========================== logs ===========================
    waiting for locator('#theia\\:menubar .p-MenuBar-itemLabel').locator('text=View') to be visible
    ============================================================

       at ../../src/theia-main-menu.ts:47

      45 |
      46 |     protected menuBarItem(label = ''): Promise<ElementHandle<SVGElement | HTMLElement> | null> {
    > 47 |         return this.page.waitForSelector(this.menuBarItemSelector(label));
         |                          ^
      48 |     }
      49 |
      50 |     protected menuBarItemSelector(label = ''): string {

        at TheiaMenuBar.menuBarItem (/Users/mac/Work/eclipsesource_theia/examples/playwright/src/theia-main-menu.ts:47:26)
        at TheiaMenuBar.openMenu (/Users/mac/Work/eclipsesource_theia/examples/playwright/src/theia-main-menu.ts:30:40)
        at /Users/mac/Work/eclipsesource_theia/examples/playwright/src/tests/theia-electron-app.test.ts:48:34

  Slow test file: ../../src/tests/theia-electron-app.test.ts (1.5m)
  Consider splitting slow test files to speed up parallel execution

  2 failed
    ../../src/tests/theia-electron-app.test.ts:38:9 › Theia Electron Application › open about dialog using menu 
    ../../src/tests/theia-electron-app.test.ts:47:9 › Theia Electron Application › toggle explorer view using menu 
  5 passed (1.9m)
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

inlann commented:
Does "playwright cannot access native UI elements" mean native file/folder selection dialogs such as 'Add folder to workspace' is not available?

xai commented:
> Does "playwright cannot access native UI elements" mean native file/folder selection dialogs such as 'Add folder to workspace' is not available?

Yes, correct. This is the second thing that I am currently working on. The plan is to use the plain FileDialogService instead of ElectronFileDialogservice for the Playwright tests.

xai commented:
> HTML-based menus don't seem to work on MacOS

Thank you very much for trying this on MacOS!

So using the command-line-switch switch, you end up with no menus at all? Not good. On Linux, I suddenly had two menubars, so I set the visibility to false in https://github.com/eclipsesource/theia/blob/54798d2dc3a81b8a985c318b63cb678fd0309bfa/packages/core/src/electron-main/electron-main-application.ts#L310 as well as in line 320. What happens on MacOS if that is removed again? Also, to have the complete information here, can you please look at whether a "window.titleBarStyle" preference is set in your `$HOME/.theia/settings.json` and, if so, what its value is?

xai commented on packages/core/src/electron-main/electron-main-application.ts:
```diff
@@ -305,13 +306,19 @@ export class ElectronMainApplication {
 
     async openDefaultWindow(): Promise<BrowserWindow> {
         const [uri, electronWindow] = await Promise.all([this.createWindowUri(), this.createWindow()]);
+        if (!this.useNativeWindowFrame) {
```
On Linux, I ended up with two menubars, so I got rid of the native one using this workaround. Maybe I should try and figure out why that happens instead of fixing the symptom :wink:

xai commented on examples/playwright/configs/playwright.config.ts:
```diff
@@ -26,11 +26,10 @@ const config: PlaywrightTestConfig = {
-        toMatchSnapshot: { threshold: 0.15 }
+        viewport: { width: 1920, height: 1080 },
+        trace: {
+            mode: 'retain-on-failure'
```
Good point. I will just disable this.

xai commented:
> > Does "playwright cannot access native UI elements" mean native file/folder selection dialogs such as 'Add folder to workspace' is not available?
> 
> Yes, correct. This is the second thing that I am currently working on. The plan is to use the plain FileDialogService instead of ElectronFileDialogservice for the Playwright tests.

Now the DefaultFileDialogService is used when the `--no-native-window-frame` command-line switch is provided. I also added a test case for opening a file using the File menu. While it works on Linux, there is a remaining issue on Windows with file system access when Electron is launched via Playwright. I am still investigating this.

xai commented on examples/playwright/src/theia-app-loader.ts:
```diff
@@ -0,0 +1,167 @@
+    ) { }
+
+    playwrightOptions(workspace?: TheiaWorkspace): object {
+        const executablePath = this.electronAppPath + '/node_modules/.bin/electron';
```
Thank you! There were several issues with paths on Windows. Following your suggestion on using `path` instead of modifying strings, it should now work on Windows.

xai replied:
Ecxept that now nothing works on Windows anymore, due to #12405 😬

inlann commented:
Hello
I ran this e2e test successfully in Jenkins. If anyone is interested in this, here is my pipeline:

```Jenkinsfile
pipeline {
    agent { label "linux-x64" }

    stages {
        stage('The Matrix') {
            matrix {
                agent { label "${AGENT}" }
                axes {
                    axis {
                        name 'AGENT'
                        values 'linux-x64'  // TODO: add 'win64'
                    }
                }
                stages {
                    stage('Install & Build') {
                        steps {
                            nodejs(nodeJSInstallationName: 'Node_v16') {
                                sh 'yarn all'
                                sh 'yarn electron build'
                                sh 'yarn electron rebuild'
                            }
                        }
                    }

                    stage('Test') {
                        steps {
                            nodejs(nodeJSInstallationName: 'Node_v16') {
                                sh 'xvfb-run -s "-screen 0 1024x768x24" yarn --cwd examples/playwright ui-tests -g "electron"'
                            }

                            nodejs(nodeJSInstallationName: 'Node_v16') {
                                sh 'xvfb-run -s "-screen 1 1280x720x24" yarn --cwd examples/playwright ui-tests -g "electron"'
                            }
                        }
                    }
                }
            }
        }
    }
}

```

xai commented:
Using a CLI switch to disable native elements was a bad idea by me, so I now replaced this with an environment variable THEIA_ELECTRON_DISABLE_NATIVE_ELEMENTS. As playwright electron is the only use case we are aware of so far, it is not justified to clutter the command line arguments with things not relevant to users.

xai commented:
**Current status** of this PR (I also updated the PR description):
* Integrated the previously explicit electron playwright tests into the existing tests
* Ensured that all existing tests can also be run in electron (`yarn ui-tests-electron`)
* Adjusted Github workflow to additionally test in electron

While testing, I noticed **two remaining problems** that in my opinion are **not specific** to the changes introduced by **this PR**:
* Nothing works on Windows and MacOS (#12405)
* Stability issues: some of the tests are flaky. Looks like race conditions, even when only one worker is used. I could reproduce this with the current master branch and ran some experiments. On my machine, some of the existing browser tests fail in 50 to 70% of all runs. I am not sure why the Github CI seems less affected by this so far.

xai commented on examples/playwright/src/tests/theia-status-bar.test.ts:
```diff
@@ -14,23 +14,31 @@
     });
 
+    test.afterAll(async () => {
+        await app.page.close();
```
They probably have not been closed at all, which became a problem, especially for the electron use case.

xai commented:
Local experiments with ~100 runs have shown that two UI tests are unstable when run in electron (output view and quick commands). As a workaround, I disabled them in this case (i.e., they are run with the browser UI tests but not with the electron UI tests) for now.

tsmaeder commented:
1) ..\..\src\tests\theia-explorer-view.test.ts:90:9 › Theia Explorer View › should show one folder named "sampleFolder", one named "sampleFolderCompact" and one file named "sample.txt"
2) ..\..\src\tests\theia-quick-command.test.ts:62:13 › Theia Quick Command › should show quick command palette
3) ..\..\src\tests\theia-text-editor.test.ts:60:9 › Theia Text Editor › should be visible and active after opening "sample.txt"
4) ..\..\src\tests\theia-toolbar.test.ts:75:9 › Theia Toolbar › should trigger the "Command Palette" toolbar tool as expect the command palette to open
5) ..\..\src\tests\theia-workspace.test.ts:54:9 › Theia Workspace › should be initialized with the contents of a file location
6) ..\..\src\tests\theia-workspace.test.ts:65:9 › Theia Workspace › should be initialized with the contents of multiple file locations

tsmaeder commented on examples/playwright/src/theia-workspace.ts:
```diff
@@ -46,15 +47,7 @@ export class TheiaWorkspace {
-            workspacePath = workspacePath.replace(/.:/, matchedChar => matchedChar.toLowerCase());
-        }
-        return workspacePath;
+        return path.normalize(this.workspacePath);
```
I'm not sure this is equivalent to the old code: for starters, it would not prepend a `/` before a relative path on unix, like the old code did.

xai replied:
Right, they do not behave the same. I also wondered what the original code was trying to solve and assumed it was supposed to deal with cross-platform paths, especially on Windows, and drive letters.

I also think, that the `/` before relative paths on Unix systems was an unintentional side effect in the old code. If you prefer the original version to `path.normalize()`, I'll of course revert it back.

xai replied:
Right, `path.normalize()` does not make drive letters lowercase, which seems to be at least necessary when running playwright from vscode. I reverted my change.

tsmaeder commented on packages/core/src/electron-main/electron-main-application.ts:
```diff
@@ -243,7 +247,15 @@ export class ElectronMainApplication {
             this.saveWindowState(browserWindow);
         } else {
-            console.warn(`no BrowserWindow with id: ${webContents.id}`);
+            console.warn(`no BrowserWindow wit  "theia": {
```
Typo: "wit".

xai replied:
Oops, that actually looks like a vi-accident of mine. Reverted to original line in a follow-up commit.

tsmaeder commented on packages/core/src/electron-main/electron-main-application.ts:
```diff
@@ -243,7 +247,15 @@ export class ElectronMainApplication {
         } else {
-            console.warn(`no BrowserWindow with id: ${webContents.id}`);
+            console.warn(`no BrowserWindow wit  "theia": {
+                "frontend": {
```
Why do we need this config info all of a sudden?

xai replied:
We don't need it. Reverted.

tsmaeder commented on packages/filesystem/src/electron-browser/file-dialog/electron-file-dialog-module.ts:
```diff
@@ -15,10 +15,37 @@
+        bind(ElectronFileDialogService).toSelf().inSingletonScope();
+        bind(FileDialogService).toService(ElectronFileDialogService);
+    } else {
+        bind(DefaultFileDialogService).toSelf().inSingletonScope();
```
This loos like it would be a copy of the browser case. If that is true, can we share this code in a function somewhere?

xai replied:
I moved the logic of whether native elements are used away from the module into the implementation. We now call the superclass of `ElectronFileDialogService` if native elements should not be used. This should also make it easier to move away from the env variable approach in future work.

tsmaeder commented on packages/filesystem/src/electron-browser/preload.ts:
```diff
@@ -21,10 +21,12 @@ import { ipcRenderer, contextBridge } from '@theia/core/electron-shared/electron
 const api: TheiaFilesystemAPI = {
     showOpenDialog: (options: OpenDialogOptions) => ipcRenderer.invoke(CHANNEL_SHOW_OPEN, options),
     showSaveDialog: (options: SaveDialogOptions) => ipcRenderer.invoke(CHANNEL_SHOW_SAVE, options),
+    useNativeDialogs: !('THEIA_ELECTRON_DISABLE_NATIVE_ELEMENTS' in process.env && process.env.THEIA_ELECTRON_DISABLE_NATIVE_ELEMENTS === '1')
```
This does not in the TheiaFileSystemAPI object. It is a generalized flag that applies to _all_ electron-based UI and should therefor go into the core package.

xai replied:
Sure, you are absolutely right. Moved to core and renamed to `useNativeElements` to simplify moving away from the env variable approach.

xai commented on examples/playwright/package.json:
```diff
@@ -19,8 +19,10 @@
-    "ui-tests": "yarn build && playwright test",
-    "ui-tests-headful": "yarn build && playwright test --headed",
+    "ui-tests": "yarn build && playwright test --config=./configs/playwright.config.ts",
+    "ui-tests-electron": "yarn build && USE_ELECTRON=true playwright test --config=./configs/playwright.config.ts",
```
For Electron, they are headful by default. To run them headless in Unix, external tools such as `xfvb` would be required afaik. As for Windows, I am still trying to figure out how to run Electron headless.

xai commented on examples/playwright/src/tests/theia-main-menu.test.ts:
```diff
@@ -14,21 +14,47 @@
+    let isElectron: boolean;
+
+    test.beforeAll(async ({ playwright, browser }) => {
+        isElectron = process.env.USE_ELECTRON === 'true';
```
Sure. In a follow-up commit, I extended our `TheiaApp` class a bit and simplified the initialization of `TheiaAppLoader`.

xai commented on examples/playwright/src/tests/theia-text-editor.test.ts:
```diff
@@ -14,27 +14,49 @@
         const ws = new TheiaWorkspace(['src/tests/resources/sample-files1']);
-        app = await TheiaApp.load(page, ws);
+        let args;
+        if (process.env.USE_ELECTRON === 'true') {
```
Sure, addressed in a follow-up commit. `TheiaAppLoader` now gets the `playwright` and `browser` objects as args and then does the rest by itself.

xai commented on examples/playwright/src/theia-app-loader.ts:
```diff
@@ -0,0 +1,182 @@
+        factory?: TheiaAppFactory<T>,
+        initialWorkspace?: TheiaWorkspace
+    ): Promise<T> {
+        const electronConfig = args.useElectron;
```
Sure, done in follow-up commit.

xai commented on examples/playwright/src/tests/theia-sample-app.test.ts:
```diff
@@ -35,12 +36,34 @@ class TheiaSampleApp extends TheiaApp {
 
-let app: TheiaSampleApp;
-
+// the tests in this file reuse a page to run faster and thus are executed serially
```
Right, and as long as we use `workers: 1` in the playwright config, this should be the default anyway. I'll remove the duplicated annotations.

xai commented on examples/playwright/src/theia-app.ts:
```diff
@@ -35,49 +35,40 @@ export const DefaultTheiaAppData: TheiaAppData = {
     shellSelector: '.theia-ApplicationShell'
 };
 
+export class TheiaAppMainPageObjects {
```
This class was initially created by @planger, so I assume he had a specific scenario in mind here.

I assumed that this class can be used by a custom `TheiaAppFactory` to pass existing instances of `TheiaStatusBar`, `TheiaQuickCommandPalette`, and `TheiaMenuBar` into the constructor of a new TheiaApp instance.

We do not use custom AppFactories in the tests for now, but it should provide better extensibility.

tsmaeder replied:
So if the class just holds a bunch of factory methods, why not have these factory methods in the TheaApp class and override the methods in a TheiaApp subclass if I want to change the created instance?

xai replied:
Sure! Done.

xai commented on examples/playwright/src/theia-app-loader.ts:
```diff
@@ -0,0 +1,182 @@
+
+}
+
+class TheiaElectronAppLoader {
```
The only reason here to argue for classes is that it might allow for better (project-specific) extensibility by keeping the generics, such as in `static async load<T extends TheiaApp>(...): Promise<T>`, which allows us to use generic `TheiaAppFactories`.

tsmaeder replied:
I don't really get the argument: creating a different app class would be done by passing in a different app factory, no? You can still keep the generics

xai replied:
You're right! Done.

xai commented on examples/playwright/src/theia-app-loader.ts:
```diff
@@ -0,0 +1,182 @@
+    }
+}
+
+export class ElectronLaunchOptions {
```
Sure! I removed the class and moved the conversion into `TheiaElectronAppLoader`.

xai commented:
Rebased to current master because of #13040.

xai commented:
Just rebased again because of a conflict.

Current state:
- Implemented suggested changes from @tsmaeder 
- Reverted unnecessary changes
- Removed electron playwright workflow in the CI for now.
  This will be re-added once we are sure that the stability is sufficient.

As discussed with @planger, we currently consider our support for electron playwright **experimental** (Playwright itself also [considers its support for Electron automation experimental](https://playwright.dev/docs/api/class-electron)). This also means that Windows is not supported at this time, as we still observe plenty of stability issues there.

planger commented:
Thanks for your review @tsmaeder! We should certainly look into the Windows support in near future.
If you are ok to merge, can you please approve this PR?
Thank you!

xai commented:
Rebased because of failed license check.
